In [42]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [43]:
train = pd.read_csv('/kaggle/input/competitions/playground-series-s6e3/train.csv')
test = pd.read_csv('/kaggle/input/competitions/playground-series-s6e3/test.csv')

In [44]:
 train.head()

,id,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,0,Male,0,Yes,Yes,29,Yes,No,DSL,Yes,...,Yes,Yes,No,No,One year,Yes,Mailed check,60.10,1653.85,No
1,1,Male,0,Yes,Yes,58,Yes,No,DSL,Yes,...,No,Yes,Yes,No,Two year,No,Credit card (automatic),69.50,3778.20,No
2,2,Male,0,Yes,No,58,Yes,Yes,Fiber optic,No,...,No,No,Yes,Yes,Month-to-month,Yes,Electronic check,100.40,5841.35,No
3,3,Female,0,No,No,1,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,69.70,70.70,Yes
4,4,Female,0,No,No,1,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.45,70.45,Yes


In [45]:
train.drop(['id'], axis = 1, inplace = True)
test_ids = test['id'].copy()
test.drop(['id'], axis = 1, inplace = True)

In [46]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 594194 entries, 0 to 594193
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   gender            594194 non-null  object 
 1   SeniorCitizen     594194 non-null  int64  
 2   Partner           594194 non-null  object 
 3   Dependents        594194 non-null  object 
 4   tenure            594194 non-null  int64  
 5   PhoneService      594194 non-null  object 
 6   MultipleLines     594194 non-null  object 
 7   InternetService   594194 non-null  object 
 8   OnlineSecurity    594194 non-null  object 
 9   OnlineBackup      594194 non-null  object 
 10  DeviceProtection  594194 non-null  object 
 11  TechSupport       594194 non-null  object 
 12  StreamingTV       594194 non-null  object 
 13  StreamingMovies   594194 non-null  object 
 14  Contract          594194 non-null  object 
 15  PaperlessBilling  594194 non-null  object 
 16  PaymentMethod     59

In [47]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 254655 entries, 0 to 254654
Data columns (total 19 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   gender            254655 non-null  object 
 1   SeniorCitizen     254655 non-null  int64  
 2   Partner           254655 non-null  object 
 3   Dependents        254655 non-null  object 
 4   tenure            254655 non-null  int64  
 5   PhoneService      254655 non-null  object 
 6   MultipleLines     254655 non-null  object 
 7   InternetService   254655 non-null  object 
 8   OnlineSecurity    254655 non-null  object 
 9   OnlineBackup      254655 non-null  object 
 10  DeviceProtection  254655 non-null  object 
 11  TechSupport       254655 non-null  object 
 12  StreamingTV       254655 non-null  object 
 13  StreamingMovies   254655 non-null  object 
 14  Contract          254655 non-null  object 
 15  PaperlessBilling  254655 non-null  object 
 16  PaymentMethod     25

In [48]:
train

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,Male,0,Yes,Yes,29,Yes,No,DSL,Yes,No,Yes,Yes,No,No,One year,Yes,Mailed check,60.10,1653.85,No
1,Male,0,Yes,Yes,58,Yes,No,DSL,Yes,Yes,No,Yes,Yes,No,Two year,No,Credit card (automatic),69.50,3778.20,No
2,Male,0,Yes,No,58,Yes,Yes,Fiber optic,No,Yes,No,No,Yes,Yes,Month-to-month,Yes,Electronic check,100.40,5841.35,No
3,Female,0,No,No,1,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,69.70,70.70,Yes
4,Female,0,No,No,1,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.45,70.45,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
594189,Male,0,No,No,57,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Two year,No,Bank transfer (automatic),97.55,5460.70,No
594190,Female,0,No,No,72,Yes,Yes,DSL,Yes,Yes,Yes,Yes,Yes,Yes,Two year,No,Bank transfer (automatic),91.95,6782.15,No
594191,Female,0,Yes,No,72,Yes,Yes,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Credit card (automatic),24.40,1871.90,No
594192,Female,0,No,No,32,Yes,Yes,Fiber optic,No,No,No,No,No,Yes,Month-to-month,Yes,Electronic check,86.00,2847.20,No


In [49]:
train['TotalCharges'] = pd.to_numeric(train['TotalCharges'], errors='coerce').fillna(0).astype(int)
test['TotalCharges'] = pd.to_numeric(test['TotalCharges'], errors='coerce').fillna(0).astype(int)
train['MonthlyCharges'] = pd.to_numeric(train['MonthlyCharges'], errors='coerce').fillna(0).astype(int)
test['MonthlyCharges'] = pd.to_numeric(test['MonthlyCharges'], errors='coerce').fillna(0).astype(int)

In [50]:
test

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
0,Female,0,Yes,No,72,Yes,Yes,Fiber optic,Yes,Yes,Yes,Yes,Yes,Yes,Two year,Yes,Electronic check,115,8061
1,Female,0,Yes,No,71,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Bank transfer (automatic),19,1336
2,Male,0,No,No,12,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Bank transfer (automatic),55,633
3,Male,0,Yes,Yes,71,Yes,Yes,DSL,Yes,No,Yes,Yes,Yes,Yes,Two year,No,Credit card (automatic),84,6457
4,Female,0,No,No,15,Yes,No,Fiber optic,Yes,No,No,No,Yes,Yes,Month-to-month,No,Electronic check,90,1233
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
254650,Male,0,Yes,Yes,72,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Credit card (automatic),19,1443
254651,Male,1,Yes,No,16,Yes,Yes,Fiber optic,No,No,No,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,100,1563
254652,Male,0,Yes,No,35,Yes,Yes,Fiber optic,Yes,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),105,3132
254653,Female,0,No,No,25,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,Yes,Credit card (automatic),20,511


In [51]:
train.shape

(594194, 20)

In [52]:
train.isnull().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [53]:
train['gender'].value_counts()

gender
Female    298738
Male      295456
Name: count, dtype: int64

In [54]:
train['PaymentMethod'].value_counts()
train['Contract'].value_counts()
train['PaymentMethod'].value_counts()


PaymentMethod
Electronic check             215372
Credit card (automatic)      133705
Mailed check                 123757
Bank transfer (automatic)    121360
Name: count, dtype: int64

In [55]:
train.dtypes

gender              object
SeniorCitizen        int64
Partner             object
Dependents          object
tenure               int64
PhoneService        object
MultipleLines       object
InternetService     object
OnlineSecurity      object
OnlineBackup        object
DeviceProtection    object
TechSupport         object
StreamingTV         object
StreamingMovies     object
Contract            object
PaperlessBilling    object
PaymentMethod       object
MonthlyCharges       int64
TotalCharges         int64
Churn               object
dtype: object

In [56]:
train.dtypes

gender              object
SeniorCitizen        int64
Partner             object
Dependents          object
tenure               int64
PhoneService        object
MultipleLines       object
InternetService     object
OnlineSecurity      object
OnlineBackup        object
DeviceProtection    object
TechSupport         object
StreamingTV         object
StreamingMovies     object
Contract            object
PaperlessBilling    object
PaymentMethod       object
MonthlyCharges       int64
TotalCharges         int64
Churn               object
dtype: object

In [57]:
cat_cols = train.select_dtypes(include=['object']).columns
print("Categorical columns:", cat_cols)

Categorical columns: Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'Churn'],
      dtype='object')


In [58]:
x = train.drop('Churn', axis = 1)
y = train['Churn']
x_test = test.copy()
x_test

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
0,Female,0,Yes,No,72,Yes,Yes,Fiber optic,Yes,Yes,Yes,Yes,Yes,Yes,Two year,Yes,Electronic check,115,8061
1,Female,0,Yes,No,71,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Bank transfer (automatic),19,1336
2,Male,0,No,No,12,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Bank transfer (automatic),55,633
3,Male,0,Yes,Yes,71,Yes,Yes,DSL,Yes,No,Yes,Yes,Yes,Yes,Two year,No,Credit card (automatic),84,6457
4,Female,0,No,No,15,Yes,No,Fiber optic,Yes,No,No,No,Yes,Yes,Month-to-month,No,Electronic check,90,1233
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
254650,Male,0,Yes,Yes,72,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Credit card (automatic),19,1443
254651,Male,1,Yes,No,16,Yes,Yes,Fiber optic,No,No,No,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,100,1563
254652,Male,0,Yes,No,35,Yes,Yes,Fiber optic,Yes,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),105,3132
254653,Female,0,No,No,25,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,Yes,Credit card (automatic),20,511


In [59]:
cat_cols = x.select_dtypes(include=['object']).columns
cat_cols

Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')

In [60]:
from sklearn.preprocessing import LabelEncoder

for col in cat_cols:
    le = LabelEncoder()
    x[col] = le.fit_transform(x[col])
    x_test[col] = le.transform(x_test[col])



x_test

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges
0,0,0,1,0,72,1,2,1,2,2,2,2,2,2,2,1,2,115,8061
1,0,0,1,0,71,1,0,2,1,1,1,1,1,1,2,0,0,19,1336
2,1,0,0,0,12,1,0,0,2,2,0,0,0,0,0,1,0,55,633
3,1,0,1,1,71,1,2,0,2,0,2,2,2,2,2,0,1,84,6457
4,0,0,0,0,15,1,0,1,2,0,0,0,2,2,0,0,2,90,1233
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
254650,1,0,1,1,72,1,0,2,1,1,1,1,1,1,2,0,1,19,1443
254651,1,1,1,0,16,1,2,1,0,0,0,2,2,2,0,1,2,100,1563
254652,1,0,1,0,35,1,2,1,2,0,2,2,2,2,0,1,0,105,3132
254653,0,0,0,0,25,1,0,2,1,1,1,1,1,1,2,1,1,20,511


In [61]:
x.dtypes

gender              int64
SeniorCitizen       int64
Partner             int64
Dependents          int64
tenure              int64
PhoneService        int64
MultipleLines       int64
InternetService     int64
OnlineSecurity      int64
OnlineBackup        int64
DeviceProtection    int64
TechSupport         int64
StreamingTV         int64
StreamingMovies     int64
Contract            int64
PaperlessBilling    int64
PaymentMethod       int64
MonthlyCharges      int64
TotalCharges        int64
dtype: object

In [62]:
y

0          No
1          No
2          No
3         Yes
4         Yes
         ... 
594189     No
594190     No
594191     No
594192     No
594193    Yes
Name: Churn, Length: 594194, dtype: object

In [63]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score, classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

In [64]:
X_train, X_val, y_train, y_val = train_test_split(x, y, test_size=0.2, random_state=42)
st = StandardScaler()

X_train_scaled = st.fit_transform(X_train)  
X_val_scaled = st.transform(X_val)           
x_test_scaled = st.transform(x_test) 
y = le.fit_transform(y)
y

array([0, 0, 0, ..., 0, 0, 1])

In [65]:
model = LogisticRegression(max_iter = 1000)

model.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

In [66]:
y_pred = model.predict(X_val_scaled)
accuracy = accuracy_score(y_val, y_pred)
accuracy

0.8528008482063969

In [67]:
print(classification_report(y_val,y_pred))

              precision    recall  f1-score   support

          No       0.90      0.91      0.91     91935
         Yes       0.68      0.65      0.67     26904

    accuracy                           0.85    118839
   macro avg       0.79      0.78      0.79    118839
weighted avg       0.85      0.85      0.85    118839



In [68]:

y_test_pred = model.predict_proba(x_test_scaled)[:, 1]
y_test_pred

array([3.67967738e-02, 1.30370711e-04, 1.16929629e-01, ...,
       2.68209417e-01, 5.59884398e-03, 5.40423710e-01])

In [69]:
from sklearn.model_selection import cross_val_score

In [70]:
x_scaled = st.fit_transform(x)
x_scaled


array([[ 1.00553879, -0.35888538,  0.95891421, ...,  1.37817462,
        -0.17388492, -0.35723273],
       [ 1.00553879, -0.35888538,  0.95891421, ..., -0.55579595,
         0.1157024 ,  0.54551756],
       [ 1.00553879, -0.35888538,  0.95891421, ...,  0.41118934,
         1.11316982,  1.42192878],
       ...,
       [-0.99449172, -0.35888538,  0.95891421, ..., -0.55579595,
        -1.33223419, -0.26462117],
       [-0.99449172, -0.35888538, -1.04284616, ...,  0.41118934,
         0.66270066,  0.15000673],
       [-0.99449172,  2.78640495,  0.95891421, ...,  0.41118934,
        -0.36694313, -1.01358529]])

In [71]:
score = cross_val_score(LogisticRegression() , x_scaled, y, cv = 7)
score

array([0.85278907, 0.85164635, 0.85397891, 0.85049184, 0.85321317,
       0.85150498, 0.85024268])

In [72]:
print(score.mean())

0.8519809999119551


In [73]:
submission = pd.DataFrame({
    'id': test_ids,
    'Churn': y_test_pred.round(5)
})

In [74]:
sub = submission.to_csv('Submission.csv',index = False)
sub

In [75]:
print("Pipeline complete. Test predictions saved to submission.csv")

Pipeline complete. Test predictions saved to submission.csv


**I built an end-to-end ML pipeline for Telco churn prediction using only logistic regression. I preprocessed the data by mapping binary features and one-hot encoding categorical variables, applied standard scaling, and split the data into training and validation sets. The model achieved an accuracy of 85% on the validation set, with performance evaluated using a classification report. I also performed 3-fold cross-validation, which gave an average score of 0.85 (85%). Finally, predictions on the test set were generated and saved in a submission-ready CSV with customer IDs and predicted churn value.**